### PR TITLE
feat: add update credential status endpoint

### PR DESCRIPTION
**Pull request by tplooker**
This PR adds a credential issuer API for managing an update to the status of an issued credential.

*Outstanding Questions*

- ~~How should the status property relate to the type property in the request? Are we assuming that the type sets the valid enumeration of values for status?~~ Yes
- ~~Is the credentialId the id of the issued credential?~~ Yes

### CODE REVIEW

**Review by OR13 — APPROVED** [review]

**Comment by OR13** [comment]
> * Is the credentialId the id of the issued credential?

Is the credentialId the id of the issued credential?

Lets do whatever will be most confusing for developers :)

**Comment by OR13** [comment]
> * ....Are we assuming that the type sets the valid enumeration of values for status?

IMO, Yes....

Support for a type is sufficient to convey support for the enumeration of values.... thats the whole point of creating unqiue type names :)

**Comment by OR13** [comment]
@tplooker if you can provide some sample JSON / curls that would be awesome.

**Comment by tplooker** [comment]
@OR13, can you elaborate on what you mean? Is this sufficient? https://github.com/w3c-ccg/vc-http-api/pull/110/files#diff-74dfbc770d6e21cb974f297ba8f1bc7015ece556537a6202e87b78622728a0cfR154

**Comment by OR13** [comment]
@tplooker I guess I was hoping for test suite JSON examples... but maybe better in a separate PR...

**Comment by tplooker** [comment]
Ah gotcha, yes fixtures and tests will be added after this PR.

**Comment by tplooker** [comment]
Would love some feedback from @msporny and @peacekeeper on this one, not trying to rush the process, just aware that this is time critical

**Comment by mprorock** [comment]
we are blocked by this, both functionally, and from a purpose of being able to run a cleanup and consistency pass on the YAML definition file, and on beginning to finalize documentation on a more complete version of the API that matches our bare minimum use case needs.

File: docs/vc-http-api.yml
Comment (msporny) [inline docs/vc-http-api.yml:61]:
```suggestion
          description: Malformed input.
```

File: docs/vc-http-api.yml
Comment (msporny) [inline docs/vc-http-api.yml:65]:
```suggestion
          description: Internal server error.
```

File: docs/vc-http-api.yml
Comment (msporny) [reply docs/vc-http-api.yml:61]: We really need to describe when these sorts of errors can occur.

File: docs/vc-http-api.yml
Comment (msporny) [reply docs/vc-http-api.yml:65]: We really need to describe when these sorts of errors can occur.

File: docs/vc-http-api.yml
Comment (msporny) [inline docs/vc-http-api.yml:157]:
```suggestion
          "id": "urn:uuid:45a44711-e457-4fa8-9b89-69fe0287c86a",
          "credentialStatus": [{
              "type": "RevocationList2020Status",
              "statusValue": 0
           }]
```

I'm concerned about both of these proposals for the object.

@dlongley, thoughts?

**Review by msporny — CHANGES_REQUESTED** [review]

File: docs/vc-http-api.yml
Comment (msporny) [reply docs/vc-http-api.yml:65]: What error is thrown if the status list type is not supported?

File: docs/vc-http-api.yml
Comment (tplooker) [reply docs/vc-http-api.yml:157]: @msporny agree with the functionality you are trying to enable, but how do we unambiguously refer to the different credential statuses in the array you have defined? I.e are we assuming that a credential will never be issued with more than one credential status of the same type?

File: docs/vc-http-api.yml
Comment (tplooker) [reply docs/vc-http-api.yml:65]: Agree we need more details on the error cases, will add some now, I would assume it would be a 400 because you are making a request that the server does not understand.

File: docs/vc-http-api.yml
Comment (tplooker) [reply docs/vc-http-api.yml:157]: Another approach would be to have this endpoint only update a single status associated to the credential and require the requestor to make multiple requests to update all the credential statuses?

Comment (msporny) [reply docs/vc-http-api.yml:157]: > are we assuming that a credential will never be issued with more than one credential status of the same type?

I believe so -- unless someone has a use case where they are publishing two revocation lists associated with the same credential.

> Another approach would be to have this endpoint only update a single status associated to the credential and require the requestor to make multiple requests to update all the credential statuses?

I don't think we have to do that... the current form allows you to update all credential statuses associated with the VC in one shot.

Comment (tplooker) [reply docs/vc-http-api.yml:157]: > I believe so -- unless someone has a use case where they are publishing two revocation lists associated with the same credential.

Ok, I agree at this stage it would be a strange application just pointing out that this mechanism for identifying the associated credential status to update is not 100% explicit and also that it is valid from the VC data model perspective to have multiple statues of the same type in a single credential.

> Unless I'm missing something, I don't think we have to do that... the current form allows you to update all credential statuses associated with the VC in one shot.

Agree im not against this style of bulk/multi status update just calling that out as our other option.

**Comment by tplooker** [comment]
Another outstanding question for this endpoint is should it be `credentials/status` or `credentials/updateStatus`

**Comment by OR13** [comment]
`credentials/status`

File: docs/vc-http-api.yml
Comment (OR13) [inline docs/vc-http-api.yml:44]:
```suggestion
  /credentials/status:
```

**Review by OR13 — CHANGES_REQUESTED** [review]
Thanks for this PR!

File: docs/vc-http-api.yml
Comment (OR13) [inline docs/vc-http-api.yml:49]: https://restfulapi.net/resource-naming/

note that the operation has the word "update" but the route should not... this route is an example of a `controller` style endpoint.

File: docs/vc-http-api.yml
Comment (OR13) [reply docs/vc-http-api.yml:44]: See https://github.com/w3c-ccg/vc-http-api/pull/110/files#r576914083

**Comment by tplooker** [comment]
Ready for re-review @msporny @OR13 @peacekeeper

**Comment by tplooker** [comment]
Also I have local test cases for revocation almost ready so if we can get this reviewed soon then I will file the next PR

File: docs/vc-http-api.yml
Comment (OR13) [inline docs/vc-http-api.yml:161]: feels like there should be an index in here....

updateCredentialStatus = (credentialId:string, statusIndex:number, statusFlag: boolean)

Comment (OR13) [reply docs/vc-http-api.yml:161]: @tplooker ^

Comment (tplooker) [reply docs/vc-http-api.yml:161]: I would have assumed that the `credentialId` would allow you to look that up? Having to supply it is exposing too much complexity to the client IMO

Comment (peacekeeper) [reply docs/vc-http-api.yml:161]: Requiring `status` to be a number feels very specific to RevocationList2020Status, no? Other credential status mechanisms in the future may have other concepts of status. Maybe allow status to be any string (and in the specific case of RevocationList2020Status the only valid string values are "0" and "1")?

Comment (tplooker) [reply docs/vc-http-api.yml]: +1 @peacekeeper I have updated this

Comment (tplooker) [reply docs/vc-http-api.yml]: @OR13 I have addressed @peacekeeper's query but will leave this open until you have had a chance to reply on your original query

**Review by OR13 — APPROVED** [review]

**Comment by tplooker** [comment]
@peacekeeper @msporny looking for re-reviews here PR been open 8 days, made changes yesterday as per @OR13 review and now approval, anything else from you needed before approval and merging?

File: docs/vc-http-api.yml
Comment (msporny) [inline docs/vc-http-api.yml:65]: No exclamation marks in error conditions! :P 

Also, these descriptions are not very descriptive. We need to fix this (not now, but later)

**Review by msporny — APPROVED** [review]

File: docs/vc-http-api.yml
Comment (tplooker) [reply docs/vc-http-api.yml:65]: But errors are exceptional and hence should be emphasised :) will update

**Comment by tplooker** [comment]
Multiple reviews and approvals 8 days open, merging.

**Pull request merged**
